### PR TITLE
Make `mlflow` depend on `mlflow-skinny` to prevent version mismtach

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/12426?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12426/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12426
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

The mlflow-and-mlflow-skinny co-installation issue has bitten us too many times before. This PR makes `mlflow x.y.z` depend on `mlflow-skinny x.y.z` to prevent their version mismatch. Once this PR is merged, we'll have three mlflow variants:

| Variant          | Dependencies                               | Notes                                                                                                                             |
| ---------------- | ------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------- |
| mlflow-skinny    | skinny dependencies                        | skinny is always skinny                                                                                                           |
| mlflow (dev)     | skinny dependencies + core dependencies    | mlflow for development, installable via `pip install git+https://github.com/mlflow/mlflow.git`. Doesn't depend on `mlflow-skinny` |
| mlflow (release) | `mlflow-skinny==x.y.z` + core dependencies | mlflow published on PyPI. Depends on `mlflow-skinny`                                                                              |

<details><summary>How to reproduce the error</summary>
<p>

```
docker run --rm python:3.8 bash -c "pip install mlflow==2.14.0 && pip install mlflow-skinny==2.13.2 && python -c 'import mlflow'"
```

Output

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/mlflow/__init__.py", line 109, in <module>
    from mlflow.tracing.fluent import (
  File "/usr/local/lib/python3.8/site-packages/mlflow/tracing/fluent.py", line 26, in <module>
    from mlflow.tracing.utils import (
ImportError: cannot import name 'extract_span_inputs_outputs' from 'mlflow.tracing.utils' (/usr/local/lib/python3.8/site-packages/mlflow/tracing/utils/__init__.py)
```

File structure:

```
> tree /usr/local/lib/python3.8/site-packages/mlflow/tracing
...
├── utils     # comes from 2.14.0
│   ├── __init__.py
│   ├── __pycache__
│   │   ├── __init__.cpython-38.pyc
│   │   └── search.cpython-38.pyc
│   └── search.py
└── utils.py  # comes from 2.13.2
```

</p>
</details> 

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

#### Code for manual testing

<details><summary>Open</summary>
<p>

```python
"""
Simulate the installation of mlflow that depends on mlflow-skinny
"""
import contextlib
import re
import shutil
import socket
import subprocess
import sys
import tempfile
from pathlib import Path


def get_safe_port():
    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
        sock.bind(("localhost", 0))
        return sock.getsockname()[1]


@contextlib.contextmanager
def serve_wheels():
    with tempfile.TemporaryDirectory() as tmpdir:
        subprocess.check_call(
            [
                sys.executable,
                "dev/build.py",
                "--package-type",
                "skinny",
            ],
            stdout=subprocess.DEVNULL,
        )
        mlflow_skinny_dir = Path(tmpdir) / "mlflow-skinny"
        shutil.copytree("dist", mlflow_skinny_dir)

        subprocess.check_call(
            [
                sys.executable,
                "dev/build.py",
                "--package-type",
                "release",
            ],
            stdout=subprocess.DEVNULL,
        )
        subprocess.check_call(["dev/build.sh"], stdout=subprocess.DEVNULL)
        mlflow_dir = Path(tmpdir) / "mlflow"
        shutil.copytree("dist", mlflow_dir)

        port = get_safe_port()
        with subprocess.Popen(
            [
                sys.executable,
                "-m",
                "http.server",
                str(port),
            ],
            cwd=tmpdir,
        ) as prc:
            try:
                yield f"http://localhost:{port}"
            finally:
                prc.terminate()


def main():
    version = re.search(
        r'^VERSION = "([a-z0-9\.]+)"$',
        Path("mlflow", "version.py").read_text(),
        re.MULTILINE,
    ).group(1)
    with serve_wheels() as url:
        subprocess.check_call(
            [
                sys.executable,
                "-m",
                "pip",
                "install",
                "--pre",
                f"mlflow=={version}",
                "--extra-index-url",
                url,
            ]
        )
        subprocess.check_call([sys.executable, "-m", "pip", "list"])


if __name__ == "__main__":
    main()
```

Output:

```
...
matplotlib                         3.7.5
mlflow                             2.14.2.dev0
mlflow-skinny                      2.14.2.dev0
numpy                              1.24.4
...
```

</p>
</details> 

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: pyproject.release.toml
Comment (harupy) [inline pyproject.release.toml:1]: autogenerated file. line 25 is important.

File: dev/build.py
Comment (harupy) [inline dev/build.py:1]: We have jobs in our internal repo that still uses `build.sh`. We need to migrate them to use this, and remove `build.sh`.

File: pyproject.release.toml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  "Intended Audience :: Information Technology",
+  "Topic :: Scientific/Engineering :: Artificial Intelligence",
+  "Topic :: Software Development :: Libraries :: Python Modules",
+  "License :: OSI Approved :: Apache Software License",
+  "Operating System :: OS Independent",
+  "Programming Language :: Python :: 3.8",
+]
+requires-python = ">=3.8"
+dependencies = [
+  "mlflow-skinny==2.14.2.dev0",
```
Comment (harupy) [inline pyproject.release.toml:25]: 👆👆👆

File: dev/pyproject.py
Comment (harupy) [inline dev/pyproject.py:1]: we might want to bundle these dev script as a CLI package to make it easier to share helper functions.

File: pyproject.release.toml
Comment (harupy) [inline pyproject.release.toml:1]: might want to insert a comment at the top explaining what this file is for.

File: pyproject.release.toml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  "pyarrow<16,>=4.0.0",
+  "querystring_parser<2",
+  "scikit-learn<2",
+  "scipy<2",
+  "sqlalchemy<3,>=1.4.0",
+  "waitress<4; platform_system == 'Windows'",
+]
+[[project.maintainers]]
+name = "Databricks"
+email = "mlflow-oss-maintainers@googlegroups.com "
```
Comment (harupy) [inline pyproject.release.toml:46]: trailing space. I'll keep this as a good first issue.

File: dev/pyproject.py [hunk 3]
```diff
@@ -54,20 +62,32 @@ def build(skinny: bool) -> None:
...
             "requires": ["setuptools"],
             "build-backend": "setuptools.build_meta",
         },
         "project": {
-            "name": "mlflow" if not skinny else "mlflow-skinny",
-            "version": version,
+            "name": package_name,
+            "version": package_version,
             "maintainers": [
                 {"name": "Databricks", "email": "mlflow-oss-maintainers@googlegroups.com "}
```
Comment (harupy) [inline dev/pyproject.py:92]: Spotted trailing whitespace. I'll keep this as a good first issue.

File: pyproject.release.toml (hunk 1 above)
Comment (BenWilson2) [reply pyproject.release.toml:25]: nice!

File: dev/build.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    return parser.parse_args()
+
+
+@contextlib.contextmanager
+def restore_changes():
+    try:
+        yield
+    finally:
+        subprocess.check_call(["git", "restore", ":^dev/build.py"])
+
```
Comment (BenWilson2) [inline dev/build.py:26]: Would it be helpful for diagnosing build-related issues to capture subprocess output and raise the full stack trace to stdout?

Something like:

```python
def run_subprocess(command):
    try:
        result = subprocess.run(command, capture_output=True, text=True, check=True)
        return result
    except subprocess.CalledProcessError as e:
        print(f"Command '{' '.join(command)}' failed with return code {e.returncode}")
        print(f"stdout:\n{e.stdout}")
        print(f"stderr:\n{e.stderr}")
        raise
``` 
?

Comment (harupy) [reply dev/build.py:26]: @BenWilson2 Thanks for the comment! `check_call` shows stdout and stderr by default.

File: pyproject.release.toml (hunk 1 above)
Comment (harupy) [inline pyproject.release.toml:25]: pip supports version control systems (VCS):

```
pandas @ git+https://github.com/pandas-dev/pandas`
```
but `mlflow-skinny` can't be installed using this format.

Comment (harupy) [inline pyproject.release.toml:25]: you can still install both mlflow and mlflow-skinny to cause the same problem:

```
pip install mlflow
pip install mlflow-skinny==1.2.3
```

**Review by BenWilson2 — APPROVED** [review]
LGTM! This is sufficient for protecting against nearly all reported issues (manually specifying different versions is a deliberate action that we don't have to solve for). 
Thanks Haru!

**Pull request merged**